clients/src/main/java/org/apache/kafka/clients/consumer/internals/SharedConsumerState.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    public SharedConsumerState(LogContext logContext,
+                               ConsumerMetadata metadata,
+                               SubscriptionState subscriptions,
+                               Time time,
+                               long retryBackoffMs,
+                               ApiVersions apiVersions) {
+        requireNonNull(logContext);
+        requireNonNull(metadata);
+        requireNonNull(subscriptions);
+        requireNonNull(time);
+        requireNonNull(apiVersions);
+
+        this.subscriptions = subscriptions;
+        this.offsetFetcherUtils = new OffsetFetcherUtils(
+            logContext,
+            metadata,
+            subscriptions,
+            time,
+            retryBackoffMs,
+            apiVersions
+        );
+        this.updatePositionsError = new SharedExceptionReference();
+        this.metadataError = new SharedExceptionReference();
+    }
+
+    OffsetFetcherUtils offsetFetcherUtils() {
+        return offsetFetcherUtils;
+    }
+
+    public SharedExceptionReference updatePositionsError() {
+        return updatePositionsError;
+    }
+
+    public SharedExceptionReference metadataError() {
+        return metadataError;
+    }
+
+    /**
+     * This method is used by {@code AsyncKafkaConsumer#updateFetchPositions()} to determine if it can skip
+     * the step of sending (and waiting for) a {@link CheckAndUpdatePositionsEvent}. {@code updateFetchPositions()}
+     * is in the critical path for the {@link Consumer#poll(Duration)}, and if the application thread can determine
+     * that it doesn't need to perform the {@link OffsetsRequestManager#updateFetchPositions(long)} call (via the
+     * {@link CheckAndUpdatePositionsEvent}), that is a big performance savings.
+     *
+     * <p/>
+     *
+     * This method performs similar checks to the start of {@link OffsetsRequestManager#updateFetchPositions(long)}:
+     *
+     * <ol>
+     *     <li>
+     *         Checks for previous exceptions during update positions
+     *         ({@code OffsetsRequestManager#cacheExceptionIfEventExpired()})
+     *     </li>
+     *     <li>
+     *         Checks that there are no positions in the {@link SubscriptionState.FetchStates#AWAIT_VALIDATION}
+     *         state ({@link OffsetFetcherUtils#getPartitionsToValidate()})
+     *     </li>
+     *     <li>
+     *         Checks that all positions are in the {@link SubscriptionState.FetchStates#FETCHING} state
+     *         ({@link SubscriptionState#hasAllFetchPositions()})
+     *     </li>
+     * </ol>
+     *
+     * If the first check fails, an exception will be thrown. If any of the second, third, or fourth checks fail, this
+     * method will return {@code false}. Otherwise, this method will return {@code true}, which signals to the
+     * application thread that the {@link CheckAndUpdatePositionsEvent} can be skipped.
+     *
+     * @return true if all checks pass, false if either of the latter two checks fail
```

Comment:
I think an important part of this method's behaviour is that it will throw exceptions in several situations. That's probably relevant in the javadoc comment on this method.

Reply:
I've added verbiage related to the exceptions to the `canSkipUpdateFetchPositions()` method JavaDoc.